### PR TITLE
ENH: TLS/SSL support in peer forwarder

### PR DESCRIPTION
**Pull request by chenqi0805**
*Issue #, if available:*

*Description of changes:*
- Add TLS/SSL params into peerforwarder config
- Adjust PeerClientPool to accommodate TLS/SSL for client creation
- Test cases to cover the above.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java [hunk 1]
```diff
@@ -3,22 +3,51 @@
...
+import static org.mockito.Mockito.verify;
+
+@RunWith(MockitoJUnitRunner.class)
 public class PeerForwarderConfigTest {
+    private static final String VALID_SSL_KEY_CERT_FILE = PeerForwarderConfigTest.class.getClassLoader()
+            .getResource("test-crt.crt").getFile();
+    private static final String INVALID_SSL_KEY_CERT_FILE = "";
+    private static List<String> TEST_ENDPOINTS = Arrays.asList("172.0.0.1", "172.0.0.2");
+
+    private static final MockedStatic<PeerClientPool> mockedPeerClientPoolClass = Mockito.mockStatic(PeerClientPool.class);
```
Comment (wrijeff) [inline data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java:32]: This line might be causing problems - per [the MockedStatic docs](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/MockedStatic.html):
>The static mock is released when this object's ScopedMock.close() method is invoked. If this object is never closed, the static mock will remain active on the initiating thread. It is therefore recommended to create this object within a try-with-resources statement unless when managed explicitly, for example by using a JUnit rule or extension.

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java [hunk 2]
```diff
@@ -28,4 +66,35 @@ public void testBuildConfig() {
...
+        settings.put(PeerForwarderConfig.STATIC_ENDPOINTS, TEST_ENDPOINTS);
+        settings.put(PeerForwarderConfig.SSL, true);
+
+        settings.put(PeerForwarderConfig.SSL_KEY_CERT_FILE, INVALID_SSL_KEY_CERT_FILE);
+        Assert.assertThrows(IllegalArgumentException.class, () -> {
+            PeerForwarderConfig.buildConfig(new PluginSetting("peer_forwarder", settings));
+        });
+
+        settings.put(PeerForwarderConfig.SSL_KEY_CERT_FILE, null);
+        Assert.assertThrows(IllegalArgumentException.class, () -> {
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java:83]: So the reason I assertThrows inline is to make sure this specific function call throws the expected Exception.

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPool.java [hunk 3]
```diff
@@ -32,8 +46,16 @@ public void setClientTimeoutSeconds(int clientTimeoutSeconds) {
 
     private TraceServiceGrpc.TraceServiceBlockingStub createGRPCClient(final String ipAddress) {
         // TODO: replace hardcoded port with customization
-        return Clients.builder(String.format("gproto+http://%s:21890/", ipAddress))
-                .writeTimeout(Duration.ofSeconds(clientTimeoutSeconds))
-                .build(TraceServiceGrpc.TraceServiceBlockingStub.class);
+        final ClientBuilder clientBuilder;
+        if (ssl) {
+            clientBuilder = Clients.builder(String.format("%s://%s:21890/", GRPC_HTTPS, ipAddress))
```
Comment (AustinTag) [inline data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPool.java:51]: Nit: I have seen people advocate for using StringBuilder in these situations. Not a blocker at all.

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfig.java [hunk 4]
```diff
@@ -37,6 +43,20 @@ public static PeerForwarderConfig buildConfig(final PluginSetting pluginSetting)
...
         final PeerClientPool peerClientPool = PeerClientPool.getInstance();
         peerClientPool.setClientTimeoutSeconds(3);
+        final boolean ssl = pluginSetting.getBooleanOrDefault(SSL, DEFAULT_SSL);
+        final String sslKeyCertChainFilePath = pluginSetting.getStringOrDefault(SSL_KEY_CERT_FILE, null);
+        final File sslKeyCertChainFile;
+        if (ssl) {
+            if (sslKeyCertChainFilePath == null || sslKeyCertChainFilePath.isEmpty()) {
+                throw new IllegalArgumentException(String.format("%s is enable, %s can not be empty or null", SSL, SSL_KEY_CERT_FILE));
+            } else {
+                sslKeyCertChainFile = new File(sslKeyCertChainFilePath);
```
Comment (AustinTag) [inline data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfig.java:53]: Nit: May want to add another OR branch in the above if statement to handle cases where the File cannot be found.

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPoolTest.java [hunk 5]
```diff
@@ -16,4 +30,36 @@ public void testGetClientValidAddress() {
 
         assertNotNull(client);
     }
+
+    @Test
+    public void testGetClientWithSSL() {
```
Comment (AustinTag) [inline data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPoolTest.java:35]: Is there no test for client without ssl?

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (AustinTag) [reply data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java:83]: ^ You can't use "expected" to assert multiple exceptions being thrown.

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfig.java (hunk 4 above)
Comment (chenqi0805) [reply data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfig.java:53]: Sure.

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPoolTest.java (hunk 5 above)
Comment (chenqi0805) [reply data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPoolTest.java:35]: testGetClientValidAddress is w/o ssl by default. But there is no connect to server test on that. Will add one for completeness.

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPool.java (hunk 3 above)
Comment (chenqi0805) [reply data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerClientPool.java:51]: Looks like StringBuilder is going to take more lines. Will revisit when the port is also customized.

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wrijeff) [reply data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/processor/peerforwarder/PeerForwarderConfigTest.java:83]: Oh, that's something I missed (didn't realize multiple exceptions here). In that case we might want to split these up into separate tests, one for an invalid file and one for null.

**Review by AustinTag — APPROVED** [review]

**Comment by chenqi0805** [comment]
There is flaky test case:
```
com.amazon.dataprepper.plugins.processor.peerforwarder.PeerClientPoolTest > testGetClientWithoutSSL FAILED
    io.grpc.StatusRuntimeException at PeerClientPoolTest.java:54
        Caused by: com.linecorp.armeria.client.UnprocessedRequestException at UnprocessedRequestException.java:45
            Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException
                Caused by: java.net.ConnectException at Errors.java:124
```
will remove for now as it might be due to using the same port for SSL and non-ssl.

**Pull request merged**